File: Java/CTCI/src/linked_list/TestLinkedListT.java
package linked_list;

import static org.junit.Assert.assertEquals;
import static org.junit.Assert.assertFalse;
import static org.junit.Assert.assertNotNull;
import static org.junit.Assert.assertTrue;

import org.junit.Test;

public class TestLinkedListT {
	
	@Test
	public void testLinkedListTInit() {
		LinkedListT ll = new LinkedListT();
		assertEquals(ll.getLength(), 0);
		assertFalse(ll.contains(5));
	}
	
	@Test 
	public void testLinkedListTInsert() {
		LinkedListT ll = new LinkedListT();
		for (int i = 0; i < 20; i++) {
			ll.prepend(i);
			assertEquals(ll.getLength(), i+1);
		}
		for (int i = 0; i < 20; i++) {
			assertTrue(ll.contains(i));
		}
		assertFalse(ll.contains(20));
		assertFalse(ll.contains(21));
		assertFalse(ll.contains(-1));
	}
	
	@Test
	public void testLinkedListTIsPalindromeEven() {
		LinkedListT ll = new LinkedListT();
		for (int i = 0; i < 3; i++) {
			ll.prepend(i);
		}
		for (int i = 2; i >= 0; i--) {
			ll.prepend(i);
		}
		assertTrue(ll.isPalindrome());
	}
	
	@Test
	public void testLinkedListTIsPalindromeOdd() {
		LinkedListT ll = new LinkedListT();
		for (int i = 0; i < 3; i++) {
			ll.prepend(i);
		}
		for (int i = 3; i >= 0; i--) {
			ll.prepend(i);
		}
		assertTrue(ll.isPalindrome());
	}
	
	@Test
	public void testLinkedListTKthLastNode() {
		LinkedListT ll = new LinkedListT(10);

		for (int i = 1; i<= 10; i++) {
			Node retval = ll.kth_last_node(i);
			assertNotNull(retval);
			assertEquals(retval.data, 10 - i);
		}
	}
	
	@Test(expected = IllegalArgumentException.class)
	public void testLinkedListTKthLastNodeIllegalArg() throws IllegalArgumentException {
		LinkedListT ll = new LinkedListT(10);
		Node retval = ll.kth_last_node(0);
	}
	
	@Test(expected = IllegalArgumentException.class)
	public void testLinkedListTKthLastNodeIllegalArg2() throws IllegalArgumentException {
		LinkedListT ll = new LinkedListT(10);
		Node retval = ll.kth_last_node(-1);
	}
	
	@Test(expected = IndexOutOfBoundsException.class)
	public void testLinkedListTKthIndexOutOfBounds() throws IndexOutOfBoundsException {
		LinkedListT ll = new LinkedListT(10);
		Node retval = ll.kth_last_node(11);	
	}
	
	@Test(expected = IndexOutOfBoundsException.class)
	public void testLinkedListTKthIndexOutOfBounds2() throws IndexOutOfBoundsException {
		LinkedListT ll = new LinkedListT(10);
		Node retval = ll.kth_last_node(12);
	}	
	
	@Test
	public void testLinkedListTRemoveDup() {
		LinkedListT ll = new LinkedListT(10, true);
		ll.remove_dups();
		assertFalse(ll.containsDuplicate());
	}
	
	@Test
	public void testLinkedListTRemoveDupNoChange() {
		LinkedListT ll = new LinkedListT(5);
		LinkedListT soln = new LinkedListT(5);
		ll.remove_dups();
		assertTrue(ll.compare(soln));
	}
	
	@Test
	public void testLinkedListTRemoveDup2() {
		LinkedListT ll = new LinkedListT(10, true);
		assertEquals(ll.getLength(), 20);
		ll.remove_dups2();
		assertEquals(ll.getLength(), 10);
		assertFalse(ll.containsDuplicate());

	}
	
	@Test
	public void testLinkedListTRemoveDup2NoChange() {
		LinkedListT ll = new LinkedListT(5);
		LinkedListT soln = new LinkedListT(5);
		ll.remove_dups2();
		assertTrue(ll.compare(soln));
	}
	
	@Test
	public void testLinkedListTPartitionNoChange() {
		LinkedListT ll = new LinkedListT(10);
		assertEquals(ll.getLength(), 10);
		LinkedListT soln = ll;
		ll.partition(20);
		assertEquals(ll.getLength(), 10);
		assertTrue(ll.compare(soln));
	}
	
	@Test
	public void testLinkedListTPartition() {
		LinkedListT ll = new LinkedListT(10);
		assertEquals(ll.getLength(), 10);
		int k = 5;
		ll.partition(5);
		assertEquals(ll.getLength(), 10);
		
		boolean partition_found = false;
		int val;
		for (int i = 0; i < ll.getLength(); i++) {
			
			val = ll.get(i);
			
			if (val >= k)
				partition_found = true;
			
			if (!partition_found) {
				assertTrue(val < k);
			} else {
				assertTrue(val >= k);
			}
		}
	}
	
	@Test(expected = IllegalArgumentException.class)
	public void testLinkedListTGetIllegalArg() throws IllegalArgumentException {
		LinkedListT ll = new LinkedListT(10);
		int retval = ll.get(-2);
	}
	
	@Test(expected = IllegalArgumentException.class)
	public void testLinkedListTGetIllegalArg2() throws IllegalArgumentException {
		LinkedListT ll = new LinkedListT(10);
		int retval = ll.get(-1);
	}
	
	@Test(expected = IndexOutOfBoundsException.class)
	public void testLinkedListTGetOutofBounds() throws IndexOutOfBoundsException {
		LinkedListT ll = new LinkedListT(10);
		int retval = ll.get(11);	
	}
	
	@Test 
	public void testLinkedListLoopDetection() {
		LinkedListT LLT1 = new LinkedListT(10, 11);	
		Node retval = LLT1.loop_detection();
		assertNotNull(retval);
	}
	
	@Test
	public void testLinkedListLoopDetectionNoLoop() {
		LinkedListT LLT1 = new LinkedListT(10);
		Node retval = LLT1.loop_detection();
		assertTrue(retval == null);
	}
	
	@Test
	public void testIsIntersecting() {

		LinkedListT ll1 = new LinkedListT(10);
		LinkedListT ll2 = new LinkedListT(1);
		LinkedListT ll3 = new LinkedListT(3);
		Node n = ll1.kth_last_node(5);
		assertNotNull(n);
		ll2.setTail(n);
		
		Node ret = LinkedList.isIntersecting(ll1, ll2);
		assertNotNull(ret);
		assertEquals(ret.data, 5);
		
		ret = LinkedList.isIntersecting(ll3, ll2);
		assertTrue(ret == null);
	}
	
	@Test
	public void testLinkedListTRemoveMiddleNode() {
		LinkedListT ll1 = new LinkedListT(10);
		Node n = ll1.kth_last_node(5);
		assertEquals(ll1.getLength(), 10);
		assertNotNull(n);
		ll1.remove_middle_node(n);
		assertEquals(ll1.getLength(), 9);
		assertFalse(ll1.contains(5));
		for (int i = 0; i < 10; i++) {
			if (i == 5)
				continue;
			assertTrue(ll1.contains(i));
		}
	}
	
	@Test
	public void testLinkedListTSumLists() {
		LinkedListT l1 = new LinkedListT(3);
		LinkedListT l2 = new LinkedListT(4);
		LinkedListT soln = new LinkedListT();
		soln.append(0);
		soln.append(2);
		soln.append(4);
		soln.append(3);
		LinkedListT sum = LinkedListT.sumLists(l1, l2);
		assertTrue(soln.compare(sum));
	}
	
	@Test
	public void testLinkedListTSumListsCarry() {
		int[] val1 = {3, 5, 7};
		int[] val2 = {6, 6, 6};
		int[] val_s = {9, 1, 4, 1};
		LinkedListT l1 = new LinkedListT(3, val1);
		LinkedListT l2 = new LinkedListT(3, val2);
		LinkedListT soln = new LinkedListT(4, val_s);
		
		LinkedListT sum = LinkedListT.sumLists(l1, l2);
		assertTrue(soln.compare(sum));
	}
}

File: Python/02_Linked_List/test_LinkedListT.py
import unittest
from LinkedListT import LinkedListT, sum_lists
from LinkedListModule import is_intersecting

class TestLinkedListTModule(unittest.TestCase):

    def test_LinkedListT_init(self):
        ll = LinkedListT()
        self.assertEqual(ll.length, 0)

    def test_LinkedListT_prepend(self):
        ll = LinkedListT()
        ll.prepend(5)
        self.assertEqual(ll.length, 1)
        self.assertTrue(ll.contains(5))
        ll.prepend(1)
        ll.prepend(2)
        self.assertEqual(ll.length, 3)
        self.assertTrue(ll.contains(1))
        self.assertTrue(ll.contains(2))
        self.assertFalse(ll.contains(0))

    def test_LinkedListT_append(self):
        ll = LinkedListT()
        ll.append(5)
        self.assertEqual(ll.length, 1)
        self.assertTrue(ll.contains(5))
        ll.append(1)
        ll.append(2)
        self.assertEqual(ll.length, 3)
        self.assertTrue(ll.contains(1))
        self.assertTrue(ll.contains(2))
        self.assertFalse(ll.contains(0))


    def test_LinkedListT_remove(self):
        ll = LinkedListT()
        ll.prepend(5)
        self.assertEqual(ll.length, 1)
        self.assertTrue(ll.contains(5))
        ll.remove(5)
        self.assertEqual(ll.length, 0)
        self.assertFalse(ll.contains(5))
        ll.prepend(1)
        ll.prepend(2)
        self.assertEqual(ll.length, 2)
        self.assertFalse(ll.contains(5))
        self.assertTrue(ll.contains(1))
        self.assertTrue(ll.contains(2))
        ll.remove(1)
        ll.remove(2)
        self.assertEqual(ll.length, 0)
        self.assertFalse(ll.contains(1))
        self.assertFalse(ll.contains(2))

    def test_LinkedListT_compare(self):
        l1 = LinkedListT(list_length=10)
        l2 = LinkedListT(list_length=10)
        self.assertTrue(l1.compare(l2))

    def test_LinkedListT_remove_dup(self):
        l1 = LinkedListT(list_length=10)
        for i in range(0, 5):
            l1.append(i)
        soln = LinkedListT(list_length=10)
        l1.remove_duplicates()
        self.assertTrue(l1.compare(soln))
        self.assertTrue(l1.length, 10)

    def test_LinkedListT_remove_dup2(self):
        l1 = LinkedListT(list_length=10)
        for i in range(0, 5):
            l1.append(i)
        soln = LinkedListT(list_length=10)
        l1.remove_duplicates2()
        self.assertTrue(l1.compare(soln))
        self.assertTrue(l1.length, 10)

    def test_LinkedListT_kth_last_node(self):
        ll = LinkedListT(list_length=10)
        for i in range(1, 10):
            success, node = ll.kth_last_node(i)
            self.assertTrue(success)
            self.assertEqual(node.data, 10-i)

    def test_LinkedListT_partition(self):
        ll = LinkedListT(list_length=10)
        for i in range(1, 9):
            ll.partition(i)
            self.assertTrue(ll.test_partition(i))

    def test_LinkedListT_is_palindrome(self):
        ll = LinkedListT(2)
        self.assertFalse(ll.isPalindrome())

        ll = LinkedListT()
        for i in range(0,3):
            ll.prepend(i)
        for i in range(2, -1, -1):
            ll.prepend(i)
        self.assertTrue(ll.isPalindrome())

        lt = LinkedListT()
        for i in range(0,3):
            lt.prepend(i)
            lt.append(i)
        self.assertTrue(lt.isPalindrome())

    def test_sum_lists(self):
        LLT1 = LinkedListT(3) # 0 1 2
        LLT2 = LinkedListT(4) # 0 1 2 3
        soln = LinkedListT()  # 0 2 4 3
        soln.append(0)
        soln.append(2)
        soln.append(4)
        soln.append(3)
        lt_sum = sum_lists(LLT1, LLT2)
        self.assertTrue(lt_sum.compare(soln))

    def test_sum_lists2(self):
        LLT1 = LinkedListT(list_length=3, init_vals=[6, 6, 6]) #
        LLT2 = LinkedListT(list_length=3, init_vals=[3, 5, 7]) # 
        soln = LinkedListT(list_length=4, init_vals=[9, 1, 4, 1])  # 0 2 4 3
        lt_sum = sum_lists(LLT1, LLT2)
        self.assertTrue(lt_sum.compare(soln))

    def test_sum_lists_carry(self):
        LLT1 = LinkedListT(8) 
        rev = LLT1.reverse()  # 7 6 5 4 3 2 1 0
        LLT2 = LinkedListT(6) 
        LLT2.remove(0)
        LLT2.remove(1)
        LLT2.remove(2)        # 3 4 5
        soln = LinkedListT()  # 0 1 1 5 3 2 1 0
        soln.append(0)
        soln.append(1)
        soln.append(1)
        soln.append(5)
        soln.append(3)
        soln.append(2)
        soln.append(1)
        soln.append(0)
        lt_sum = sum_lists(rev, LLT2)
        self.assertTrue(lt_sum.compare(soln))

    def test_is_circular(self):
        LT = LinkedListT(10, circular=True)
        isloop, node = LT.loop_detection()
        self.assertTrue(isloop)
        self.assertEqual(node.data, 0)

        LT2 = LinkedListT(10, circular=False)
        isloop, node = LT2.loop_detection()
        self.assertFalse(isloop)

        LT3 = LinkedListT(10, circular=True, k=3)
        
        isloop, node = LT3.loop_detection()
        self.assertTrue(isloop)
        self.assertEqual(node.data, 3)

    def test_is_intersecting_false(self):
        ll1 = LinkedListT(10)
        ll2 = LinkedListT(4)
        ret, node = is_intersecting(ll1, ll2) 
        self.assertFalse(ret)
        ret, node = is_intersecting(ll2, ll1) # communitative
        self.assertFalse(ret)

    def test_is_intersecting_beg(self):
        ll1 = LinkedListT(10)
        ll2 = LinkedListT(4)
        ll2.append(50)
        ret, node = ll1.kth_last_node(10)
        self.assertTrue(ret)
        ll2.setTail(node)
        ret, node = is_intersecting(ll1, ll2)
        self.assertTrue(ret)
        self.assertEqual(node.data, 0)
        ret, node = is_intersecting(ll2, ll1) # commutative
        self.assertTrue(ret)
        self.assertEqual(node.data, 0)

    def test_is_intersecting_mid(self):
        ll1 = LinkedListT(10)
        ll2 = LinkedListT(4)
        ret, node = ll1.kth_last_node(3)
        self.assertTrue(ret)
        ll2.setTail(node)
        ret, node = is_intersecting(ll1, ll2)
        self.assertTrue(ret)
        self.assertEqual(node.data, 7)
        ret, node = is_intersecting(ll2, ll1) # commutative
        self.assertTrue(ret)
        self.assertEqual(node.data, 7)

    def test_is_intersecting_end(self):
        ll1 = LinkedListT(10)
        ll2 = LinkedListT(4)
        ret, node = ll1.kth_last_node(1)
        self.assertTrue(ret)
        ll2.setTail(node)
        ret, node = is_intersecting(ll1, ll2)
        self.assertTrue(ret)
        self.assertEqual(node.data, 9)
        ret, node = is_intersecting(ll2, ll1) # commutative
        self.assertTrue(ret)
        self.assertEqual(node.data, 9)

if __name__ == '__main__':
    unittest.main()
